teps/0141-platform-context-variables.md

```diff
@@ -221,190 +232,319 @@ The downside of this approach is that Tasks can't easily reference the values of
-- Are we coupling Tekton and other projects (e.g. Knative, Sigstore) in this proposal?
-- What is the impact of the coupling to operators e.g. maintenance & end-to-end testing?
-- Are there opinionated choices being made in this proposal? If so, are they necessary and can users extend it with
-their own choices?
--->
+This proposal avoids opinionation about which parameters should be injected by a platform and what their names should be; for example, it doesn't require an "eventID" or "repoURL" even if these parameters are needed for many use cases. Pipelines does not have any dependencies on Chains as a result of this proposal, although Chains will need to be updated to be able to interpret the "context" field.
 
 ### Conformance
 
-<!--
-https://github.com/tektoncd/community/blob/main/design-principles.md#conformance
-
-- Does this proposal require the user to understand how the Tekton API is implemented?
-- Does this proposal introduce additional Kubernetes concepts into the API? If so, is this necessary?
-- If the API is changing as a result of this proposal, what updates are needed to the
-[API spec](https://github.com/tektoncd/pipeline/blob/main/docs/api-spec.md)?
--->
+This proposal supports Tekton conformance, as it helps platform builders avoid adding non-conformant syntax to PipelineRuns in order to inject parameters. Platforms may choose what parameters to support, and supporting platform context variables will not be required for conformance.
 
 ### User Experience
 
-<!--
-(optional)
+This TEP centers around the needs of the ["platform builder" user profile](../user-profiles.md#3-platform-builder). There are multiple ways platform builders can interact with ["Pipeline and Task users](../user-profiles.md#2-pipeline-and-task-users), impacting the syntax we choose for referencing platform-provided parameters. There are also multiple ways platform builders can implement separation of concerns, impacting the model we choose for performing RBAC on who may set the "context" field.
 
-Consideration about the user experience. Depending on the area of change,
-users may be Task and Pipeline editors, they may trigger TaskRuns and
-PipelineRuns or they may be responsible for monitoring the execution of runs,
-via CLI, dashboard or a monitoring system.
+#### Example: Pipelines as Code
 
-Consider including folks that also work on CLI and dashboard.
--->
+With [Pipelines as Code](https://pipelinesascode.com/), a cluster operator is responsible for [installation](https://pipelinesascode.com/docs/install/operator_installation/) and configuring [connected repositories](https://pipelinesascode.com/docs/guide/repositorycrd/). This might be an infrastructure team member setting up CI/CD for their organization. Team members (CI/CD users) can then write PipelineRuns [referencing context from the connected repository](https://pipelinesascode.com/docs/guide/authoringprs/), and Pipelines as Code is responsible for creating the PipelineRun.
```

Comment:
We can support this in Pipelines as Code for the simple key/value  we provide (ie: standard params as per screenshot) 

<img width="739" alt="image" src="https://github.com/tektoncd/community/assets/98980/4ed925ee-8af8-4262-9029-1f6699ff9d9e">

We do have advanced use case we are introducing in the form of CEL expressions inside the expressions

https://github.com/openshift-pipelines/pipelines-as-code/pull/1466

which i'd think would be useful for platform variable as well.. altho may need an API change we want to consider here (i.e: a `type` field which in this case would be cel and evaluate as CEL expression) 

We need as well being able to access those variables replaced from anywhere, I think there is some limitation currently where the params substitution occur on pipelinerun?

We probably want to specify what happen after remote resolver resolution occured, for example for PaaC we need to be able to do this  when we are using [the remote pipeline feature](https://pipelinesascode.com/docs/guide/resolver/#remote-pipeline-annotations) or [remote task ](https://pipelinesascode.com/docs/guide/resolver/#remote-task-annotations) if there is $(context.platforms.*) variables they would be expanded after resolution..

for the foreseeable future we would have to support both form from our side but if we get features parity with platform context variable we would try to slowly make users migrate to it by generating context variable from our {{ }} variables

let me know if you want me to expand on any of those point since i am guessing it's a comment with a lot of different topics.

Reply:
Thanks for the suggestions and information! 

> We do have advanced use case we are introducing in the form of CEL expressions inside the expressions

Happy to add CEL if this proposal makes sense!

> We need as well being able to access those variables replaced from anywhere, I think there is some limitation currently where the params substitution occur on pipelinerun?

Yes, this tep proposes that users can access those variables anywhere, the same as current `params`. I don't think we want the limitation to only use in pipelinerun

> We probably want to specify what happen after remote resolver resolution occured

After the resolution, those $(context.platforms.*) variables would be replaced. I think we mentioned this in the tep's design details section.

> for the foreseeable future we would have to support both form from our side but if we get features parity with platform context variable we would try to slowly make users migrate to it by generating context variable from our {{ }} variables

It's totally up to the vendors who want to adopt this feature. For vendors who don't have pipeline-as-code, they can use this feature to build their own params populate logic. I don't know if it makes sense for pac users to migrate to this feature. Since it is also gated by feature flag, and the api is omit empty, it won't affect any current users if it is disabled.